src/components/NewStack.js

```diff
@@ -126,24 +127,36 @@ class NewStack extends PureComponent {
 
   onSubmit (e) {
     e.preventDefault()
+    console.log(e)
 
     let updateOperationsState = false
 
-    const operations = this.state.operations.map((operation) => {
-      const required = this.props.operations[operation.name].required || []
-      const addedOptions = Object.keys(operation.options)
-      const missing = required.filter((field) => {
-        const index = addedOptions.indexOf(field)
-        return index === -1 || !operation.options[field].length
-      })
+    const ajv = new Ajv({
+      allErrors: true
+    })
 
-      operation.errors = {}
-      missing.forEach((field) => {
-        operation.errors[field] = `${field} is required`
-      })
+    this.operationValidators = {}
+    Object.keys(this.props.operations).forEach(key => {
+      const operation = this.props.operations[key]
+      if (key === 'noop' && Array.isArray(operation.properties)) {
+        operation.properties = {}
+      }
+      this.operationValidators[key] = ajv.compile(operation)
+    })
 
-      if (missing.length) {
+    const operations = this.state.operations.map(operation => {
+      this.operationValidators[operation.name](operation.options)
+      if (!this.operationValidators[operation.name](operation.options)) {
+        console.log(this.operationValidators[operation.name].errors)
         updateOperationsState = true
+
+        this.operationValidators[operation.name].errors.forEach((e) => {
+          console.log(e.dataPath)
+          operation.errors[e.dataPath.replace('.', '')] = e.message
+          if (e.keyword === 'required') {
+            operation.errors[e.params.missingProperty] = e.message
+          }
```

Comment:
probably the normal case (`operation.errors[e.dataPath.replace('.', '')] = e.message`) needs to be in the else here, no?